File: src/main/java/com/freya02/botcommands/internal/ExecutableInteractionInfo.java
package com.freya02.botcommands.internal;

import com.freya02.botcommands.internal.application.CommandParameter;
import com.freya02.botcommands.internal.runner.MethodRunner;
import kotlin.reflect.KFunction;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Collectors;

public interface ExecutableInteractionInfo {
	KFunction<?> getMethod();

	@NotNull
	MethodRunner getMethodRunner();

	@NotNull
	MethodParameters<? extends CommandParameter<?>> getParameters();

	@NotNull
	default List<? extends CommandParameter<?>> getOptionParameters() {
		return getParameters().stream().filter(CommandParameter::isOption).collect(Collectors.toList());
	}

	@NotNull
	Object getInstance();
}

File: src/main/java/com/freya02/botcommands/internal/ExecutableInteractionInfo.kt
package com.freya02.botcommands.internal

import com.freya02.botcommands.internal.parameters.MethodParameter
import com.freya02.botcommands.internal.runner.MethodRunner
import kotlin.reflect.KFunction

interface ExecutableInteractionInfo {
    val method: KFunction<*>
    val methodRunner: MethodRunner
    val parameters: MethodParameters
    val optionParameters: List<MethodParameter>
        get() = parameters.filter { it.isOption }
    val instance: Any
}